internal/journal/volumegroupjournal.go

```diff
@@ -0,0 +1,429 @@
+		journalPool,
+		reqName,
+		namePrefix string) (*VolumeGroupData, error)
+	UndoReservation(
+		ctx context.Context,
+		csiJournalPool,
+		snapshotGroupName,
+		reqName string) error
+	GetVolumeGroupAttributes(
+		ctx context.Context,
+		pool,
+		objectUUID string) (*VolumeGroupAttributes, error)
+	ReserveName(
+		ctx context.Context,
+		journalPool string,
+		journalPoolID int64,
+		reqName,
+		namePrefix string) (string, string, error)
+	AddVolumeSnapshotMapping(
+		ctx context.Context,
+		pool,
+		reservedUUID,
+		volumeID,
+		snapshotID string) error
+	RemoveVolumeSnapshotMapping(
+		ctx context.Context,
+		pool,
+		reservedUUID,
+		volumeID string) error
+}
+
+// VolumeGroupJournalConfig contains the configuration and connection details.
+type VolumeGroupJournalConfig struct {
+	*Config
+	*Connection
+}
+
+// NewCSIVolumeroupJournal returns an instance of VolumeGroupJournal for groups.
+func NewCSIVolumeroupJournal(suffix string) VolumeGroupJournal {
+	return &VolumeGroupJournalConfig{
+		Config: &Config{
+			csiDirectory:            "csi.vol.groups." + suffix,
+			csiNameKeyPrefix:        "csi.vol.group",
+			cephUUIDDirectoryPrefix: "csi.vol.group.",
+			csiNameKey:              "csi.groupname",
+			namespace:               "",
+		},
+	}
+}
+
+// SetNamespace sets the namespace for the journal.
+func (sgj *VolumeGroupJournalConfig) setNamespace(ns string) {
+	sgj.Config.namespace = ns
+}
+
+// NewCSIVolumeGroupJournalWithNamespace returns an instance of VolumeGroupJournal for
+// volume groups using a predetermined namespace value.
+func NewCSIVolumeGroupJournalWithNamespace(suffix, ns string) VolumeGroupJournal {
+	j := NewCSIVolumeroupJournal(suffix)
+	j.setNamespace(ns)
+
+	return j
+}
+
+// Connect establishes a new connection to a ceph cluster for journal metadata.
+func (sgj *VolumeGroupJournalConfig) Connect(
+	monitors,
+	namespace string,
+	cr *util.Credentials,
+) (*VolumeGroupJournalConfig, error) {
+	conn, err := sgj.Config.Connect(monitors, namespace, cr)
+	if err != nil {
+		return nil, err
+	}
+	sgj.Connection = conn
+
+	return sgj, nil
+}
+
+// Destroy frees any resources and invalidates the journal connection.
+func (sgj *VolumeGroupJournalConfig) Destroy() {
+	sgj.Connection.Destroy()
+}
+
+// VolumeGroupData contains the GroupUUID and VolumeGroupAttributes for a
+// volume group.
+type VolumeGroupData struct {
+	GroupUUID             string
+	VolumeGroupAttributes *VolumeGroupAttributes
+}
+
+func generateVolumeGroupName(namePrefix, groupUUID string) string {
+	if namePrefix == "" {
+		namePrefix = defaultVolumeGroupNamingPrefix
+	}
+
+	return namePrefix + groupUUID
+}
+
+/*
+CheckReservation checks if given request name contains a valid reservation
+  - If there is a valid reservation, then the corresponding VolumeGroupData for
+    the snapshot group is returned
+  - If there is a reservation that is stale (or not fully cleaned up), it is
+    garbage collected using the UndoReservation call, as appropriate
```

Comment:
It is unclear to me if after garbage collecting, it creates a new reservation?

Reply:
No, its undo the reservation, its caller responsibility to create new reservation, or a new reservation is created in the next RPC call as the reservation is deleted.